x/exp/batch/batch.go

```diff
@@ -0,0 +1,387 @@
+//
+//  1. When a Permit Policy Condition refers to an ignored value, the Condition is dropped from the Policy.
+//  2. When a Forbid Policy Condition refers to an ignored value, the Policy is dropped.
+//  3. When a Scope clause refers to an ignored value, that scope clause is set to match any.
+//
+// Errors may be returned for a variety of reasons:
+//
+//   - It will error in case of a context error.
+//   - It will error in case any of PARC are an incorrect type at authorization.
+//   - It will error in case there are unbound variables.
+//   - It will error in case there are unused variables.
+//
+// The result passed to the callback must be used / cloned immediately and not modified.
+func Authorize(ctx context.Context, ps *cedar.PolicySet, entityMap types.Entities, request Request, cb Callback) error {
+	be := &batchEvaler{}
+	var found []types.String
+	found = findVariables(request.Principal, found)
+	found = findVariables(request.Action, found)
+	found = findVariables(request.Resource, found)
+	found = findVariables(request.Context, found)
+	for _, key := range found {
+		if _, ok := request.Variables[key]; !ok {
+			return fmt.Errorf("%w: %v", errUnboundVariable, key)
+		}
+	}
+	for k := range request.Variables {
+		if !slices.Contains(found, k) {
+			return fmt.Errorf("%w: %v", errUnusedVariable, k)
+		}
+	}
+	for _, vs := range request.Variables {
+		if len(vs) == 0 {
+			return nil
+		}
+	}
+	pm := ps.Map()
+	be.policies = make([]idPolicy, len(pm))
+	i := 0
+	for k, p := range pm {
+		be.policies[i] = idPolicy{PolicyID: k, Policy: (*ast.Policy)(p.AST())}
+		i++
+	}
+	be.callback = cb
+	switch {
+	case request.Principal == nil:
+		return fmt.Errorf("%w: principal", errMissingPart)
+	case request.Action == nil:
+		return fmt.Errorf("%w: action", errMissingPart)
+	case request.Resource == nil:
+		return fmt.Errorf("%w: resource", errMissingPart)
+	case request.Context == nil:
+		return fmt.Errorf("%w: context", errMissingPart)
+	}
+	be.env = eval.InitEnv(&eval.Env{
+		Entities:  entityMap,
+		Principal: request.Principal,
+		Action:    request.Action,
+		Resource:  request.Resource,
+		Context:   request.Context,
+	})
+	be.Values = Values{}
+	for k, v := range request.Variables {
+		be.Variables = append(be.Variables, variableItem{Key: k, Values: v})
+	}
+	slices.SortFunc(be.Variables, func(a, b variableItem) int {
+		return len(a.Values) - len(b.Values)
+	})
+
+	// resolve ignores if no variables exist
+	if len(be.Variables) == 0 {
+		doPartial(be)
+		fixIgnores(be)
+	}
+
+	return doBatch(ctx, be)
+}
+
+func doPartial(be *batchEvaler) {
+	var np []idPolicy
+	for _, p := range be.policies {
+		part, keep := eval.PartialPolicy(be.env, p.Policy)
+		if !keep {
+			continue
+		}
+		np = append(np, idPolicy{PolicyID: p.PolicyID, Policy: part})
+	}
+	be.compiled = false
+	be.policies = np
+	be.evalers = nil
+}
+
+func fixIgnores(be *batchEvaler) {
```

Comment:
I'm trying to understand why we need this. Is it just for the PARC that's passed in the callback in the case of ignores?

When I removed this code and the checks that the PAR are entities and the C is a context from `diagnosticAuthzWithCallback()`, there was only one test (`TestBatch.ignoreContext`) that failed in a way I thought might be significant, but I actually don't understand the failure:

It says that this result is missing from the result set:
```
{Request:{Principal:P::"1" Action:A::"1" Resource:R::"1" Context:{}} Values:map[resource:R::"1"] Decision:deny Diagnostic:{Reasons:[] Errors:[]}}
```

But here's the result set:
```
[
    {Request:{Principal:P::"1" Action:A::"1" Resource:R::"1" Context:{}} Values:map[resource:R::"1"] Decision:deny Diagnostic:{Reasons:[] Errors:[]}}
    {Request:{Principal:P::"1" Action:A::"1" Resource:R::"2" Context:{}} Values:map[resource:R::"2"] Decision:allow Diagnostic:{Reasons:[{PolicyID:0 Position:{Filename: Offset:0 Line:0 Column:0}}] Errors:[]}}
]
```

You can see it matches the first element exactly. Is this an empty map vs nil map thing or something?

Reply:
Let's talk about this one.

Reply:
Added comment, thanks!